lib/puppet/provider/openldap_schema/olc.rb

```diff
@@ -0,0 +1,91 @@
+require 'tempfile'
+
+Puppet::Type.type(:openldap_schema).provide(:olc) do
+
+  # TODO: Use ruby bindings (can't find one that support IPC)
+
+  defaultfor :osfamily => :debian, :osfamily => :redhat
+
+  commands :slapcat => 'slapcat', :ldapmodify => 'ldapmodify', :slaptest => 'slaptest'
+
+  mk_resource_methods
+
+  @@openldap_schema_paths = []
+
+  def self.instances
+    i = []
+    slapcat(
+      '-b',
+      'cn=config',
+      '-H',
+      'ldap:///???(objectClass=olcSchemaConfig)'
+    ).split("\n\n").collect do |paragraph|
+      name = nil
+      paragraph.split("\n").collect do |line|
+        case line
+        when /^cn: /
+          i << new(
+            :ensure => :present,
+            :name   => line.match(/^cn: (\{\d+\})?(.*)$/).captures[1]
+          )
+        end
+      end
+    end
+    i
+  end
+
+  def self.prefetch(resources)
+    mods = instances
+    missing = false
+	# we need to add all schemas again to add a single missing schema (can't compile with missing references)
+    resources.keys.each do |name|
+        unless provider = mods.find{ |mod| mod.name == name }
+            missing = true
+        end
+    end
+
+    unless missing
+        resources.keys.each do |name|
+          if provider = mods.find{ |mod| mod.name == name }
+            resources[name].provider = provider
+          end
+        end
+    end
+  end
+
+  def create
+    if resource[:path]
+      @@openldap_schema_paths.push(resource[:path])
```

Comment:
The slaptest utility needs to have all the schemas in the config file that the schema could depend on, as far as I know you cannot use the already installed schemas. By not matching the found items with the given list puppet is forced to create all resources again so we get a list of all schemas in the correct order.
I don't think this will help for inter-dependent schemas as the slaptest utility stops processing the import statements as soon as it misses a dependency.